pkg/monitor/shutdown/consumer.go

```diff
@@ -0,0 +1,160 @@
+package shutdown
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	Reason                       = "GracefulShutdownWindow"
+	KubeAPIServerShutdownLocator = "shutdown/apiserver"
+)
+
+var (
+	namespaceToServer = map[string]string{
+		"openshift-kube-apiserver": "kube-apiserver",
+		"openshift-apiserver":      "openshift-apiserver",
+		// TODO: oauth apiserver does not seem to emit any shutdown events
+		"openshift-oauth-apiserver": "oauth-apiserver",
+	}
+)
+
+// Monitor abstracts the monitor API
+type Monitor interface {
+	StartInterval(t time.Time, condition monitorapi.Condition) int
+	EndInterval(startedInterval int, t time.Time)
+}
+
+func newConsumer(monitor Monitor) *consumer {
+	return &consumer{
+		monitor:   monitor,
+		byHost:    make(map[string]*corev1.Event),
+		processed: make(map[types.UID]struct{}),
+	}
+}
+
+type consumer struct {
+	monitor   Monitor
+	lock      sync.Mutex
+	processed map[types.UID]struct{}
+	byHost    map[string]*corev1.Event
+}
+
+func (c *consumer) Consume(event *corev1.Event) {
+	if event == nil || !interesting(event) {
+		return
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var ok bool
+	if _, ok := c.processed[event.UID]; ok {
+		return
+	}
+
+	key := key(event)
+	var start, end *corev1.Event
+	start, ok = c.byHost[key]
```

Comment:
Any chance that the start event could be missing and then mess up the logic (treating previous end as new start)?

Reply:
I added some comments to make it more clear, let me know if it helps